File: src/services/mediaService.ts
import { supabase } from './supabaseClient';
import type { MediaItem } from '../types';

/**
 * Fetch A&F products from Supabase
 */
export async function fetchProducts(): Promise<MediaItem[]> {
  const { data, error } = await supabase
    .from('products')
    .select('*')
    .order('id', { ascending: true });

  if (error) {
    console.error('Error fetching products from Supabase:', error);
    throw error;
  }

  return data || [];
}

/**
 * Fetch user media from Supabase
 */
export async function fetchUserMedia(userId: string): Promise<MediaItem[]> {
  const { data, error } = await supabase
    .from('user_media')
    .select('*')
    .eq('user_id', userId)
    .order('created_at', { ascending: false });

  if (error) {
    console.error('Error fetching user media from Supabase:', error);
    throw error;
  }

  return data || [];
}

/**
 * Upload a user photo to Supabase Storage
 */
export async function uploadUserPhoto(
  file: File,
  userId: string
): Promise<{ url: string; error: Error | null }> {
  try {
    const fileExt = file.name.split('.').pop();
    const fileName = `${Date.now()}.${fileExt}`;
    const filePath = `${userId}/${fileName}`;

    // Upload to Supabase Storage
    const { error: uploadError } = await supabase.storage
      .from('user-images')
      .upload(filePath, file);

    if (uploadError) {
      throw uploadError;
    }

    // Get public URL
    const { data: urlData } = supabase.storage
      .from('user-images')
      .getPublicUrl(filePath);

    const publicUrl = urlData.publicUrl;

    // Insert metadata into user_media table
    const { error: insertError } = await supabase
      .from('user_media')
      .insert({
        user_id: userId,
        url: publicUrl,
        type: 'image',
      });

    if (insertError) {
      throw insertError;
    }

    return { url: publicUrl, error: null };
  } catch (error) {
    console.error('Error uploading photo:', error);
    return { url: '', error: error as Error };
  }
}

/**
 * Fetch an authenticated image from Supabase Storage
 * Returns a blob URL that can be used in <img> src
 */
export async function fetchAuthenticatedImage(url: string): Promise<string> {
  try {
    // Extract the file path from the Supabase storage URL
    // URL format: https://{project}.supabase.co/storage/v1/object/public/user-images/{path}
    const urlParts = url.split('/user-images/');
    if (urlParts.length !== 2) {
      throw new Error('Invalid Supabase storage URL format');
    }

    const filePath = urlParts[1];

    // Use Supabase's download method which handles authentication automatically
    const { data, error } = await supabase.storage
      .from('user-images')
      .download(filePath);

    if (error) {
      throw error;
    }

    if (!data) {
      throw new Error('No data returned from storage');
    }

    // Create blob URL from the downloaded blob
    const blobUrl = URL.createObjectURL(data);
    return blobUrl;
  } catch (error) {
    console.error('Error fetching authenticated image:', error);
    throw error;
  }
}

File: ios/Aura/Services/MediaService.swift
import Foundation
import PhotosUI
import SwiftUI
import Supabase

final class MediaService {
    static let shared = MediaService()
    private let supabase = SupabaseService.shared.client

    private init() {}

    // MARK: - Fetch

    func fetchProducts() async throws -> [MediaItem] {
        try await supabase
            .from("products")
            .select()
            .order("id", ascending: true)
            .execute()
            .value
    }

    func fetchUserMedia(userId: String) async throws -> [MediaItem] {
        try await supabase
            .from("user_media")
            .select()
            .eq("user_id", value: userId)
            .order("created_at", ascending: false)
            .execute()
            .value
    }

    // MARK: - Upload

    /// Uploads a photo selected from PhotosPicker and inserts metadata into user_media.
    func uploadPhoto(item: PhotosPickerItem, userId: String) async throws -> String {
        guard let data = try await item.loadTransferable(type: Data.self) else {
            throw UploadError.dataLoadFailed
        }

        guard data.count <= 5 * 1024 * 1024 else {
            throw UploadError.fileTooLarge
        }

        let fileName = "\(Int(Date().timeIntervalSince1970)).jpg"
        let filePath = "\(userId)/\(fileName)"

        _ = try await supabase.storage
            .from("user-images")
            .upload(filePath, data: data, options: FileOptions(contentType: "image/jpeg"))

        let publicURL = try supabase.storage
            .from("user-images")
            .getPublicURL(path: filePath)

        struct UserMediaInsert: Encodable {
            let user_id: String
            let url: String
            let type: String
        }

        try await supabase
            .from("user_media")
            .insert(UserMediaInsert(user_id: userId, url: publicURL.absoluteString, type: "image"))
            .execute()

        return publicURL.absoluteString
    }

    // MARK: - Authenticated Image URL

    /// Returns a 1-hour signed URL for authenticated Supabase Storage images.
    func signedURL(for url: String) async throws -> URL {
        guard let filePath = url.components(separatedBy: "/user-images/").last else {
            throw UploadError.invalidURL
        }
        return try await supabase.storage
            .from("user-images")
            .createSignedURL(path: filePath, expiresIn: 3600)
    }

    // MARK: - Errors

    enum UploadError: LocalizedError {
        case dataLoadFailed
        case fileTooLarge
        case invalidURL

        var errorDescription: String? {
            switch self {
            case .dataLoadFailed: return "Could not load image data."
            case .fileTooLarge: return "Image must be smaller than 5 MB."
            case .invalidURL: return "Invalid storage URL."
            }
        }
    }
}
